ape_hardhat/__init__.py

```diff
@@ -0,0 +1,237 @@
+try:
+    from importlib.metadata import PackageNotFoundError as _PackageNotFoundError  # type: ignore
+    from importlib.metadata import version as _version  # type: ignore
+except ModuleNotFoundError:
+    from importlib_metadata import PackageNotFoundError as _PackageNotFoundError  # type: ignore
+    from importlib_metadata import version as _version  # type: ignore
+
+try:
+    __version__ = _version(__name__)
+except _PackageNotFoundError:
+    # package is not installed
+    __version__ = "<unknown>"
+
+import atexit
+import ctypes
+import platform
+import random
+import signal
+import sys
+import time
+from subprocess import PIPE, Popen, call
+from typing import Any, List, Optional
+
+from ape import plugins
+from ape_http.providers import DEFAULT_SETTINGS, EthereumProvider, NetworkConfig
+
+EPHEMERAL_PORTS_START = 49152
+EPHEMERAL_PORTS_END = 60999
+HARDHAT_CHAIN_ID = 31337
+HARDHAT_CONFIG = """
+// See https://hardhat.org/config/ for config options.
+module.exports = {
+};
+"""
+HARDHAT_START_NETWORK_RETRIES = [0.1, 0.2, 0.3, 0.5, 1.0]  # seconds between network retries
+HARDHAT_START_PROCESS_ATTEMPTS = 3  # number of attempts to start subprocess before giving up
+
+
+def _signal_handler(signum, frame):
+    """Runs on SIGTERM and SIGINT to force ``atexit`` handlers to run."""
+    atexit._run_exitfuncs()
+    sys.exit(0)
+
+
+def _set_death_signal():
+    """Automatically sends SIGTERM to child subprocesses when parent process dies."""
+    if platform.uname().system == "Linux":
+        # from: https://stackoverflow.com/a/43152455/75956
+        # the first argument "1" is PR_SET_PDEATHSIG
+        # the second argument is what signal to send the child
+        libc = ctypes.CDLL("libc.so.6")
+        return libc.prctl(1, signal.SIGTERM)
+
+
+class HardhatSubprocessError(RuntimeError):
+    pass
+
+
+class HardhatNetworkConfig(NetworkConfig):
+    # --fork <URL, JSON-RPC server to fork from>
+    fork_url: Optional[str] = None
+
+    # --fork-block-number <INT, block number to fork from>
+    fork_block_number: Optional[int] = None
+
+    # --port <INT, default from Hardhat is 8545, but our default is to assign a random port number>
+    port: Optional[int] = None
+
+    # retry strategy configs, try increasing these if you're getting HardhatSubprocessError
+    network_retries: List[float] = HARDHAT_START_NETWORK_RETRIES
+    process_attempts: int = HARDHAT_START_PROCESS_ATTEMPTS
+
+
+class HardhatProvider(EthereumProvider):
+    def __post_init__(self):
+        self._hardhat_web3 = (
+            None  # we need to maintain a separate per-instance web3 client for Hardhat
+        )
+        self.port = None
+        self.process = None
+
+        hardhat_config_file = self.network.config_manager.PROJECT_FOLDER / "hardhat.config.js"
+
+        # Write the hardhat config file in the Ape project dir if it doesn't exist yet
```

Comment:
Some of these comments are a bit distracting. I think the code is readable enough without some of these, like this one and the one for checking `npx` / `hardhat` installations.

Reply:
Ok, removed those two comments and another that was redundant.